crates/in_stream/src/mem.rs

```diff
@@ -0,0 +1,433 @@
+use crate::*;
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    io::{Error, ErrorKind, Result},
+};
+use url2::prelude::*;
+
+const SCHEME: &'static str = "mem";
+const PORT: u16 = 4242;
+
+#[derive(Debug)]
+/// memory connection specific bind config
+pub struct MemBindConfig {}
+
+impl Default for MemBindConfig {
+    fn default() -> Self {
+        Self {}
+    }
+}
+
+impl InStreamConfig for MemBindConfig {}
+
+/// bind to a virtual in-process ipc "interface"
+#[derive(Debug)]
+pub struct InStreamListenerMem {
+    url: Url2,
+    recv: crossbeam_channel::Receiver<InStreamMem>,
+    accept_queue: Vec<InStreamMem>,
+}
+
+impl InStreamListenerMem {
+    pub fn bind(url: &Url2, config: MemBindConfig) -> Result<Self> {
+        InStreamListenerMem::raw_bind(url, config)
+    }
+
+    /// private constructor, you probably want `bind`
+    fn priv_new(url: Url2, recv: crossbeam_channel::Receiver<InStreamMem>) -> Self {
+        Self {
+            url,
+            recv,
+            accept_queue: Vec::new(),
+        }
+    }
+}
+
+impl Drop for InStreamListenerMem {
+    fn drop(&mut self) {
+        get_mem_manager().unbind(&self.url);
+    }
+}
+
+impl InStreamListener<&mut [u8], &[u8]> for InStreamListenerMem {
+    type Stream = InStreamMem;
+
+    fn raw_bind<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self> {
+        let _ = MemBindConfig::from_gen(config)?;
+        get_mem_manager().bind(url)
+    }
+
+    fn binding(&self) -> Url2 {
+        self.url.clone()
+    }
+
+    fn accept(&mut self) -> Result<<Self as InStreamListener<&mut [u8], &[u8]>>::Stream> {
+        loop {
+            // first, drain all pending connections from our recv channel
+            match self.recv.try_recv() {
+                Ok(stream) => {
+                    self.accept_queue.push(stream);
+                }
+                Err(crossbeam_channel::TryRecvError::Empty) => break,
+                Err(crossbeam_channel::TryRecvError::Disconnected) => {
+                    // wait until our user has accepted all pending connections
+                    // before letting them know the channel is broken
+                    if self.accept_queue.is_empty() {
+                        return Err(ErrorKind::BrokenPipe.into());
+                    }
+                }
+            }
+        }
+        if self.accept_queue.is_empty() {
+            // acceptor is non-blocking we have nothing to return
+            return Err(Error::with_would_block());
+        }
+        // pull the next item off the queue
+
+        Ok(self.accept_queue.remove(0))
+    }
+}
```

Comment:
I'm not in love this with non atomic pattern of accessing the vector (`is_empty / remove(0)`) but as it's currently a single threaded engine there's nothing technically wrong with it.

Reply:
yeah... alas `remove_item` on vec isn't stable... and rust's borrow rules make this safe even if we were using multiple threads... if accept_queue was an Arc<Mutex<_>> we'd just have to make sure the lock guard was in scope for both operations, right?